Improve Logical Simplification Handling in sympy.logic.boolalg

sahinfalcon opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #25451 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

- Revised the `binary_check_and_simplify` function in the `boolalg` module to handle binary symbols more effectively in logical expressions.
- Eliminated unnecessary error raises for certain relational expressions that involve binary symbols.

#### Other comments
Separately we could consider improving the detection of erroneous mixing of symbols as Expr and Boolean but I don't think it is very important: time would be better spent adding a BooleanSymbol class and considering if we would ever be able to deprecate using Symbol as a Boolean.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* logic
  * Improved handling of logical simplification involving binary symbols like `Eq`
 
<!-- END RELEASE NOTES -->

sahinfalcon committed:
logic: Improve handling of logical simplification

The current simplification of logical expressions containing
binary symbols like `Eq` results in unexpected behaviour. This
introduces changes to improve the handling of such cases.

The changes include:
- Revising the `binary_check_and_simplify` function in the `boolalg`
    module to ensure consistent handling of binary symbols in logical
    expressions.
- Removing unnecessary error raises when dealing with certain relational
    expressions that involve binary symbols.

These changes aim to enhance the behaviour of logical simplification,
particularly when working with expressions that mix Boolean and
Relational components.

This commit also addresses related test failures and provides a more
accurate handling of logical simplification.

Example:
The expression "expr = Or(And(a,c), Eq(a,b))" is no longer incorrectly
simplified to "a & c" such that:
print(expr.subs({a:True, b:True, c:False})) # returns True

sahinfalcon committed:
Update .mailmap

oscarbenjamin commented on sympy/logic/tests/test_boolalg.py:
```diff
@@ -1345,3 +1341,8 @@ def test_relational_threeterm_simplification_patterns_numerically():
+
+def test_issue_25451():
+    x = Or(And(a, c), Eq(a, b))
+    assert x != And(a, c)
```
Test with `!=` is not ideal because many possible objects would pass the test. A better test could be:
```
assert isinstance(x, Or)
assert set(x.args) == {And(a, c), Eq(a, b)}
```

sahinfalcon replied:
Thank you, I will update it now. I appreciate your input.

sahinfalcon committed:
Refine test method for issue_25451

The test method for issue_25451 has been refined to improve its accuracy
and precision based on valuable feedback received.
Instead of using the != operator, the new approach employs isinstance()
and set comparison to ensure that the structure of the expression
matches the expected components. This way, it will not be the case that
many possible objects can pass the test.

oscarbenjamin commented:
Looks good to me. Thanks!
